### PR TITLE
Adds two forms of a diff filter expression that can be used to filter and process lines.

### PR DESCRIPTION
**Pull request by jdbaldry**
**What this PR does / why we need it**:

`pdiff` returns a human-readable, colorized diff of the previous and current lines.
`diff` returns a patch file format diff of the previous and current lines.

If no diff exists, the filter function indicates that the unprocessed line should not be kept, also behaving similar to the `uniq` tool.

It was proposed that this diff functionality would allow easier construction of alerting rules on changes to consistent logged data. The patch format can be used to construct a before and after from a single log line.

The human-readable "pretty" format is more useful in tools like logcli or Grafana Explore although as the diff is directional, it is dependent on the query direction.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

The PR also attempts to document the interfaces involved based upon how I understood them to work that needs verification.
These changes are in a separate commit that can be easily dropped.

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: pkg/logql/log/diff.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	pretty bool
+	// prev is the previous line.
+	prev []byte
+}
+
+// Filter implements Filterer.
+// The line is only kept if there is a difference between the current line and the
+// previous line.
+// The first line passed to the expression is necessarily always kept.
+func (df *DiffFilter) Filter(line []byte) (keep bool) {
```
Comment (owen-d) [inline pkg/logql/log/diff.go:23]: Wouldn't it be more efficient to just use `[]byte == []byte` to test if there's a diff? I think the only time you need to _compute_ the diff is when it exists, which can be more efficiently checked by equality testing.

File: pkg/logql/ast.go [hunk 2]
```diff
@@ -1171,3 +1173,31 @@ func (e *LabelReplaceExpr) String() string {
...
+
+func newDiffFilterExpr(op string) *DiffFilterExpr {
+	return &DiffFilterExpr{
+		filter: log.NewDiffFilter(op == OpPrettyDiff),
+		pretty: op == OpPrettyDiff,
+	}
+}
+
+// TODO(jdb): is it shardable?
+func (e *DiffFilterExpr) Shardable() bool { return false }
```
Comment (owen-d) [inline pkg/logql/ast.go:1191]: This should be shardable :)

**Review by owen-d — COMMENTED** [review]
Cool addition! I've left some comments and I'd love to see some documentation here as well.

File: pkg/logql/log/diff.go (hunk 1 above)
Comment (jdbaldry) [reply pkg/logql/log/diff.go:23]: Good point, will make that change.

**Comment by jdbaldry** [comment]
Absolutely needs some documentation. If the overall addition is reasonable in its current form ("pdiff" and "diff") expressions, I'll get started on that as soon as I can :) 

Thanks for taking a look so quickly.

**Pull request closed without merging**